lib/stages/applyGains.cpp

```diff
@@ -363,7 +376,15 @@ void applyGains::fetch_thread() {
         }
 
         // update gains
-        state_id_t state_id = dm.create_state<gainState>(update_id, transition_interval).first;
+        state_id_t state_id;
+        if (new_state) {
+            state_id = dm.create_state<gainState>(update_id, transition_interval).first;
+        }
+        // Use the current dataset ID
+        else {
+            const auto update = gains_fifo.get_update(double_to_ts(new_ts)).second;
```

Comment:
update shadows a variable from the next outer scope (it confused me for about one minute :smile: )

Reply:
Oh, nice catch! I did not realise that.

Reply:
Actually this has me thinking, will this line actually get the last update added to `gains_fifo`? As I am passing `new_ts` to `get_update()`.

Reply:
> Actually this has me thinking, will this line actually get the last update added to `gains_fifo`? As I am passing `new_ts` to `get_update()`.

```
/**
     * @brief Fetch the latest update for a frame with the given timestamp.
     *
     * Finds the update from the queue that should be applied to a frame
     * with the given timestamp.
     *
     * @param  timestamp  The timestamp of a frame.
     *
     * @returns  The value of the most recent update from before the given timestamp
     *           and the timestamp associated to the update. If the queue is empty, or
     *           all updates are in the future, a nullptr is returned as update.
     */
```
It will get the last update (with `new_state == true`) that came in with a timestamp before the timestamp of the new update. I guess that makes sense :shrug: